### PR TITLE
test: add route with invalid remote_addr (E2E)

### PR DESCRIPTION
**Pull request by idbeta**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [x] E2E Testcase
- [ ] New feature provided
- [ ] Improve performance

- Related issues
resolve https://github.com/apache/apisix-dashboard/issues/631
___
### Bugfix
- Description

- How to fix?

___
### New feature or improvement
- Describe the details and related test reports.

### CODE REVIEW

**Comment by membphis** [comment]
bad title, I can not know what are you doing from title and description

File: api/test/e2e/route_e2e_remote_addr_test.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package e2e
+
+import (
+	"net/http"
+	"testing"
+	//  "time"
```
Comment (membphis) [inline api/test/e2e/route_e2e_remote_addr_test.go:22]: remove useless code, keep code tidy

File: api/test/e2e/route_e2e_remote_addr_test.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+//Teardown
+func TestConsumer_teardown(t *testing.T) {
+	_ = []HttpTestCase{
+		{
+			caseDesc:     "delete route",
+			Object:       MangerApiExpect(t),
+			Method:       http.MethodDelete,
+			Path:         "/apisix/admin/routes/r1",
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusOK,
```
Comment (membphis) [inline api/test/e2e/route_e2e_remote_addr_test.go:100]: should 404 here

**Review by membphis — CHANGES_REQUESTED** [review]

File: api/test/e2e/route_e2e_remote_addr_test.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+			Object:   MangerApiExpect(t),
+			Method:   http.MethodPut,
+			Path:     "/apisix/admin/routes/r1",
+			Body: `{
+					"uri": "/test_uri",
+					"remote_addr": "127.0.0.",
+					"upstream": {
+						"type": "roundrobin",
+						"nodes": {
+							"172.16.238.20:1980": 1
```
Comment (moonming) [inline api/test/e2e/route_e2e_remote_addr_test.go:38]: we not support this format anymore

File: api/test/e2e/route_e2e_remote_addr_test.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+func TestRoute_add_with_invalid_remote_addr(t *testing.T) {
+	tests := []HttpTestCase{
+		{
+			caseDesc: "config route with invalid remote_addr",
+			Object:   MangerApiExpect(t),
+			Method:   http.MethodPut,
+			Path:     "/apisix/admin/routes/r1",
+			Body: `{
+					"uri": "/test_uri",
+					"remote_addr": "127.0.0.",
```
Comment (moonming) [inline api/test/e2e/route_e2e_remote_addr_test.go:34]: is this test json chema?

File: api/test/e2e/route_e2e_remote_addr_test.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+					   "uri": "/test_uri",
+					   "remote_addr": "127.0.0.aa",
+					   "upstream": {
+						   "type": "roundrobin",
+						   "nodes": {
+							   "172.16.238.20:1980": 1
+						   }
+					   }
+				   }`,
+			Headers:      map[string]string{"Authorization": token},
```
Comment (moonming) [inline api/test/e2e/route_e2e_remote_addr_test.go:60]: not check data plane

**Review by moonming — CHANGES_REQUESTED** [review]

File: api/test/e2e/route_e2e_remote_addr_test.go (hunk 4 above)
Comment (idbeta) [reply api/test/e2e/route_e2e_remote_addr_test.go:34]: as caseDesc said, this is test for `add route with invalid remote_addr`

File: api/test/e2e/route_e2e_remote_addr_test.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+ * limitations under the License.
+ */
+package e2e
+
+import (
+	"net/http"
+	"testing"
+)
+
+//case 1: config route with invalid remote_addr
```
Comment (membphis) [inline api/test/e2e/route_e2e_remote_addr_test.go:24]: I think this comment is useless, pls remove it

File: api/test/e2e/route_e2e_remote_addr_test.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+			caseDesc:     "verify route",
+			Object:       APISIXExpect(t),
+			Method:       http.MethodGet,
+			Path:         "/test_uri",
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusNotFound,
+			Sleep:        sleepTime,
+		},
+
+		// TODO it's a bug here, see: https://github.com/apache/apisix-dashboard/issues/872
```
Comment (membphis) [inline api/test/e2e/route_e2e_remote_addr_test.go:109]: we can remove line 109 and 111 too, we have the issues

File: api/test/e2e/route_e2e_remote_addr_test.go (hunk 3 above)
Comment (idbeta) [reply api/test/e2e/route_e2e_remote_addr_test.go]: I already use the new format now.

File: api/test/e2e/route_remote_addr_test.go [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+				}`,
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusBadRequest,
+		},
+		{
+			caseDesc:     "verify route",
+			Object:       APISIXExpect(t),
+			Method:       http.MethodGet,
+			Path:         "/test_uri",
+			Headers:      map[string]string{"Authorization": token},
```
Comment (membphis) [inline api/test/e2e/route_remote_addr_test.go:49]: We can remove this line

File: api/test/e2e/route_remote_addr_test.go [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+				   }`,
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusBadRequest,
+		},
+		{
+			caseDesc:     "verify route",
+			Object:       APISIXExpect(t),
+			Method:       http.MethodGet,
+			Path:         "/test_uri",
+			Headers:      map[string]string{"Authorization": token},
```
Comment (membphis) [inline api/test/e2e/route_remote_addr_test.go:76]: ditto

File: api/test/e2e/route_remote_addr_test.go (hunk 8 above)
Comment (membphis) [reply api/test/e2e/route_remote_addr_test.go:49]: sorry,  I think we should keep this

**Review by membphis — APPROVED** [review]

**Comment by membphis** [comment]
@moonming take a look when you have time

**Comment by nic-chen** [comment]
need test cases for correct remote addr

**Comment by nic-chen** [comment]
the PR title should have a better one.

**Comment by idbeta** [comment]
> need test cases for correct remote addr

yes, you can check this PR https://github.com/apache/apisix-dashboard/pull/884

**Comment by membphis** [comment]
> the PR title should have a better one.

agree, why do you submit two PRs about `route + remote_addr`?

**Review by membphis — APPROVED** [review]

**Review by nic-chen — APPROVED** [review]

**Comment by juzhiyuan** [comment]
ping @moonming

File: api/test/e2e/route_e2e_remote_addr_test.go (hunk 4 above)
Comment (moonming) [reply api/test/e2e/route_e2e_remote_addr_test.go]: Isn't this the job of json chema? why should e2e to cover?

**Comment by moonming** [comment]
> ping @moonming

As I said days before, these checks should be the work of json chema.
Is there a problem with the json schema that can't be verified?

**Comment by idbeta** [comment]
> > ping @moonming
> 
> As I said days before, these checks should be the work of json chema.
> Is there a problem with the json schema that can't be verified?

**Comment by membphis** [comment]
> I think we should not trust json schema completely, so it is reasonable to cover it in e2e.

@moonming do you have time to review this PR again? PR was blocked now

**Comment by moonming** [comment]
no resposne from @idbeta

File: api/test/e2e/route_remote_addr_test.go [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+						   "type": "roundrobin",
+						   "nodes": [{
+								"host": "172.16.238.20",
+								"port": 1980,
+								"weight": 1
+						    }]
+					   }
+				   }`,
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusBadRequest,
```
Comment (moonming) [inline api/test/e2e/route_remote_addr_test.go:73]: we need to check the response body.

**Review by moonming — CHANGES_REQUESTED** [review]

File: api/test/e2e/route_remote_addr_test.go (hunk 10 above)
Comment (moonming) [reply api/test/e2e/route_remote_addr_test.go:73]: Otherwise, there is no way to determine whether the json schema really works

Comment (idbeta) [reply api/test/e2e/route_remote_addr_test.go:74]: fixed.

File: api/test/e2e/route_remote_addr_test.go [hunk 11]
```diff
@@ -42,6 +42,7 @@ func TestRoute_add_with_invalid_remote_addr(t *testing.T) {
 				}`,
 			Headers:      map[string]string{"Authorization": token},
 			ExpectStatus: http.StatusBadRequest,
+			ExpectBody:   "scheme validate fail",
```
Comment (moonming) [inline api/test/e2e/route_remote_addr_test.go:45]: `scheme` -> schema, `fail` -> `failed`
 Is there any specific failed reason?

Comment (idbeta) [reply api/test/e2e/route_remote_addr_test.go]: fixed.

Comment (idbeta) [reply api/test/e2e/route_remote_addr_test.go]: I have raised the issue of typo https://github.com/apache/apisix-dashboard/issues/914

File: api/test/e2e/route_e2e_remote_addr_test.go (hunk 4 above)
Comment (idbeta) [reply api/test/e2e/route_e2e_remote_addr_test.go]: I think we should not trust json schema completely, so it is reasonable to cover it in e2e.

File: api/test/e2e/route_remote_addr_test.go (hunk 11 above)
Comment (idbeta) [reply api/test/e2e/route_remote_addr_test.go]: ping @moonming

File: api/test/e2e/route_remote_addr_test.go [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+					"uri": "/hello",
+					"remote_addr": "127.0.0.",
+					"upstream": {
+						"type": "roundrobin",
+						"nodes": [{
+							"host": "172.16.238.20",
+							"port": 1980,
+							"weight": 1
+						}]
+					}
```
Comment (moonming) [inline api/test/e2e/route_remote_addr_test.go:41]: indentation

File: api/test/e2e/route_remote_addr_test.go [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+					   "uri": "/hello",
+					   "remote_addr": "127.0.0.aa",
+					   "upstream": {
+						   "type": "roundrobin",
+						   "nodes": [{
+								"host": "172.16.238.20",
+								"port": 1980,
+								"weight": 1
+						    }]
+					   }
```
Comment (moonming) [inline api/test/e2e/route_remote_addr_test.go:71]: ditto

File: api/test/e2e/route_remote_addr_test.go [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+					   "remote_addrs": ["127.0.0.1","192.168.0."],
+					   "upstream": {
+						   "type": "roundrobin",
+						   "nodes": [{
+								"host": "172.16.238.20",
+								"port": 1980,
+								"weight": 1
+						    }]
+					   }
+				   }`,
```
Comment (moonming) [inline api/test/e2e/route_remote_addr_test.go:102]: ditto

File: api/test/e2e/route_remote_addr_test.go [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+						"nodes": [{
+							"host": "172.16.238.20",
+							"port": 1980,
+							"weight": 1
+						}]
+					}
+				}`,
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusBadRequest,
+			ExpectBody:   "remote_addr: Does not match pattern",
```
Comment (moonming) [inline api/test/e2e/route_remote_addr_test.go:45]: we need to check the complete body, not part of it

File: api/test/e2e/route_remote_addr_test.go (hunk 12 above)
Comment (idbeta) [reply api/test/e2e/route_remote_addr_test.go:41]: fixed

File: api/test/e2e/route_remote_addr_test.go (hunk 13 above)
Comment (idbeta) [reply api/test/e2e/route_remote_addr_test.go]: fixed

File: api/test/e2e/route_remote_addr_test.go (hunk 14 above)
Comment (idbeta) [reply api/test/e2e/route_remote_addr_test.go]: fixed

File: api/test/e2e/route_remote_addr_test.go (hunk 15 above)
Comment (idbeta) [reply api/test/e2e/route_remote_addr_test.go]: fixed

Comment (idbeta) [reply api/test/e2e/route_remote_addr_test.go]: ping @moonming Can you review it when you have time?

**Review by moonming — APPROVED** [review]

**Pull request merged**